rust/processor/src/parquet_manager.rs

```diff
@@ -0,0 +1,554 @@
+            parquet_rust_struct_vec.as_slice().schema().unwrap().clone()
+        },
+        Some(ParquetStruct::Transaction(struct_model)) => {
+            let mut  parquet_rust_struct_vec = Vec::new();
+            parquet_rust_struct_vec.push(struct_model.clone());
+            parquet_rust_struct_vec.as_slice().schema().unwrap().clone()    
+        },
+        Some(ParquetStruct::WriteSetChange(struct_model)) => {
+            let mut  parquet_rust_struct_vec = Vec::new();
+            parquet_rust_struct_vec.push(struct_model.clone());
+            parquet_rust_struct_vec.as_slice().schema().unwrap().clone()
+        },
+        _ => {
+            error!("No schema found for the parquet struct");
+            panic!("No schema found for the parquet struct");
+        }
+    };
+    schema
+}
+
+pub async fn create_parquet_manager_loop(
+    gcs_client: Arc<GCSClient>,
+    new_gap_detector_sender: kanal::AsyncSender<ParquetProcessingResult>, 
+    parquet_struct_receivers: Vec<kanal::AsyncReceiver<ParquetData>>, 
+    processor_name: &str,
+    bucket_name: String,
+) {
+
+    let processor_name = processor_name.to_owned();
+
+    for receiver in parquet_struct_receivers {
+        let mut parquet_manager = ParquetManager::new(bucket_name.clone());
+        let gcs_client = gcs_client.clone();
+        let processor_name = processor_name.clone();
+        let new_gap_detector_sender = new_gap_detector_sender.clone();
+        // spawan a thread per struct and per struct , we know the type of the struct and we can write it to the parquet file.
+        tokio::spawn(async move {            
+            loop {
+                let txn_pb_res = match receiver.recv().await {
+                    Ok(txn_pb_res) => {
+                        info!(
+                            processor_name = processor_name.clone(),
+                            service_type = PROCESSOR_SERVICE_TYPE,
+                            "[Parquet Handler] Received transaction data from processor",
+                        );
+                        txn_pb_res
+                    },
+                    Err(_e) => {
+                        error!(
+                            processor_name = processor_name.clone(),
+                            service_type = PROCESSOR_SERVICE_TYPE,
+                            error = ?_e,
+                            "[Parquet Handler] Parquet manager channel has been closed",
+                        );
+                        ParquetData {                  // maybe add a flag that can tell it's empty
+                            data: Vec::new(),
+                            last_transaction_timestamp: None,
+                            transaction_version_to_struct_count: AHashMap::new(),
+                        }
+                    },
+                };
+
+                let result = parquet_manager.write_and_upload(txn_pb_res, &gcs_client, new_gap_detector_sender.clone()).await;
```

Comment:
instead of cloning the `new_gap_detector_sender` every time, maybe the parquet manager should have it's own copy of it?